flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/GenericArrayData.java

```diff
@@ -132,6 +135,25 @@ public boolean isNullAt(int pos) {
 		return !isPrimitiveArray && ((Object[]) array)[pos] == null;
 	}
 
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		GenericArrayData that = (GenericArrayData) o;
+		return size == that.size &&
+			isPrimitiveArray == that.isPrimitiveArray &&
+			Arrays.deepEquals(new Object[] {array}, new Object[]{that.array});
```

Comment:
We need to use `Arrays.deepEquals` for `GenericRowData.equals`?

Reply:
Yes. I think we need that. `GenericRowData` may contains `byte[]` which should use `Arrays.deepEquals`.

Reply:
I revisited all `GenericXXX` classes again for hashCode and equals.